core/src/main/java/org/jruby/util/io/OpenFile.java

```diff
@@ -1298,7 +1298,8 @@ public int fillbuf(ThreadContext context) {
                     r = readInternal(context, this, fd, rbuf.ptr, 0, rbuf.capa);
 
                     if (r < 0) {
-                        if (waitReadable(context, fd)) {
+                        if (posix.getErrno() == Errno.EAGAIN || posix.getErrno() == Errno.EWOULDBLOCK
+                                && waitReadable(context, fd)) {
```

Comment:
will this do two foreign calls for errno value?

Reply:
In this case it's not a foreign call; `posix` is a shim class that hides native vs JDK but simulates the posix style. It should be retrieved once, though.